pg_search/src/postgres/customscan/pdbscan/exec_methods/top_n.rs

```diff
@@ -270,16 +269,83 @@ impl ExecMethod for TopNScanExecState {
             (self.limit as f64 * self.scale_factor).max(self.chunk_size as f64) as usize;
         let next_offset = self.offset + local_limit;
 
-        self.search_results = state
-            .search_reader
-            .as_ref()
-            .unwrap()
-            .search_top_n_in_segments(
-                self.segments_to_query(state.search_reader.as_ref().unwrap(), state.parallel_state),
-                self.orderby_info.as_ref(),
-                local_limit,
-                self.offset,
-            );
+        // TODO: only execute aggregations on our first query.
+        let aggregations = self.prepare_aggregations(state);
+        let agg_limits = AggregationLimitsGuard::new(
+            Some(gucs::adjust_work_mem().get().try_into().unwrap()),
+            // TODO: configure?
+            Some(tantivy::aggregation::DEFAULT_BUCKET_LIMIT),
+        );
+
+        // Run the TopN (and optional aggregate) query.
+        self.search_results = if let Some(orderby_info) = self.orderby_info.as_ref() {
+            let maybe_aggregation_collector = aggregations.as_ref().map(|aggregations| {
+                // TODO: (Optionally?) wrap in MVCCFilterCollector.
+                // See also https://github.com/paradedb/paradedb/issues/3253.
+                DistributedAggregationCollector::from_aggs(
+                    aggregations.aggregations.clone(),
+                    agg_limits.clone(),
+                )
+            });
```

Comment:
I think we should re-use `ParallelAggregationWorker` here:

https://github.com/paradedb/paradedb/blob/2faaf50722cd366feca5b83ea11869f2465a1063/pg_search/src/aggregate/mod.rs#L224-L275

You can also checkout how `MVCCFilterCollector` is wrapped in there.

Reply:
> I think we should re-use ParallelAggregationWorker here:

We can't: in the context of a TopN, we are already executing inside of a parallel worker provided by the CustomScan framework. TopN already handles checking out a subset of the segments to execute using the `ParallelScanState`.

If we used a separate parallel worker, then we wouldn't actually get the advantage of executing the aggregate and the topn in the same pass.

Reply:
How about this patch? https://github.com/paradedb/paradedb/compare/moe/merged-stuhood.window-aggregates-execution...moe/mvcc-stuhood.window-aggregates-execution